### ISSUE TITLE
Enforce unique URLs

### ISSUE THREAD

**Issue by regisphilibert** [gohugoio/hugo#6928]
### Context

In order to discern control of permalink from filename, Hugo allows to use `slug` or `title` or other variable to build such URLs.

With #6310, the filename option will become less reliable and many projects will fallback to using slugs or titles or else.

### The problem

Currently, before creating a URL reference for an entry from user's permalink settings, Hugo does not check for duplicate URLs.

So with the following permalink structure:

```
permalinks:
  notifications: notiticiations/:title
```

And the following pages:

```
title: New York Traffic delayed by bridge maintenance
date: 2015-02-09T04:00:00+00:00
```

```
title: New York Traffic delayed by bridge maintenance
date: 2019-01-09T04:00:00+00:00
```

The two following entries will use the same URL (`notifications/new-york-traffic-delayed-by-bridge-maintenance`) and one will ultimately be overwritten by the other.

### Current workaround
One can of course add more variables like `:month` or `:year` this will still not be a 100% futureproof fix and might not match the project's URL structure requirements.

Editors have to use Slugs (so does the setting) to manually identify duplicate "titles" dating back years for some.

### Proposal

If it is possible for Hugo to create unique URLs and store them for later reference during the build. It should produce the two pages at:

- `notifications/new-york-traffic-delayed-by-bridge-maintenance`
- `notifications/new-york-traffic-delayed-by-bridge-maintenance-1`

**Comment by bep** [comment]
>to manually identify duplicate 

There is a flag that enables "duplicate warnings".

As to your proposal, I think it's probably a good idea. What do the others do in this situation?

A couple of things to consider:

* Assuming we don't store any state about this between builds, we need to assign these numbers serialized (1 thread) and in the same order every time. This is doable today, but I would probably want to use all 8 cores on my new MacBook once we get more content sources ...
* It would be a little bit harder to manually link to pages.

**Comment by regisphilibert** [comment]
> What do the others do in this situation?

WordPress suffixes with numbers similar to my example. The order of the index is I suppose the order of creation (row ID order maybe). 

Others just prepend with the row id when available. If, Hugo ever produces a unique ID for its page (which I guess it should when remote data comes into play) then this could also be a new permalink variable. `/blog/:slug-:id`

> It would be a little bit harder to manually link to pages.

If you're talking about `ref` and `.GetPage`, current behaviour should not be affected. A warning when ambiguous etc...

**Comment by jeremyzilar** [comment]
Would also be nice to run a report in the terminal that shows me all unique URLs and the associated files.

**Comment by Jos512** [comment]
> If it is possible for Hugo to create unique URLs and store them for later reference during the build.

I see that for your use case this is a good solution. But I'm not sure if it helps people in a 'blogging environment'.

If Hugo finds a duplicate URLs, I'd rather see a warning/info message so I can tweak it myself. If Hugo generates unique URLs automatically for me, the manual links between pages get broken.

> There is a flag that enables "duplicate warnings".

Unfortunately `--path-warnings` only shows outputs with `hugo`, not `hugo server`. That makes it a bit harder to spot duplicate URLs problems *before* production.

**Comment by ndalliard** [comment]
I think this is still a great feature request.

As @Jos512 points out, the `--path-warnings` flag doesn't work for me with `hugo server`, even though it is listed as a flag for the `hugo server` command [in the docs](https://gohugo.io/commands/hugo_server/). With the permalinks set to `/:slug/` I am not able to identify duplicate permalinks without running `hugo` (which I don't do locally, I use `hugo server` there).

**Comment by SirUli** [comment]
Actually I'd rephrase the feature request as "tell me issue with duplicate ids also in hugo server" rather than automatically generating suffixes for URLs. WordPress in comprison has a database that stores these URLs then forever whereas hugo would decide every time on "who is first". So if I then bring in another content module as a "first one" all my URLs and of course all the links might be messed up.

Best Regards, Uli

**Comment by SirUli** [comment]
Still relevant - justification is above
